File: src/utils/createWaveform.js

/**
 * Creates an array of frequency and amplitude pairs for a given waveform.
 * @param {string} waveform - The name of the waveform ("sine", "square", "triangle", "sawtooth").
 * @param {number} baseFreq - The fundamental frequency.
 * @returns {Array<[number, number]>} A 2D array of [frequency, amplitude] pairs.
 */
export const createWaveform = (waveform, baseFreq) => {
    const harmonics = [];
    const nyquist = 20000; // Use 20kHz as the upper limit

    if (baseFreq > nyquist) {
        return [[baseFreq, 1]];
    }

    switch (waveform) {
        case "sine":
            // A sine wave has only the fundamental frequency.
            harmonics.push([baseFreq, 1]);
            break;

        case "square":
            // Square wave: odd harmonics, amplitude 1/n.
            for (let n = 1; ; n += 2) {
                const freq = baseFreq * n;
                if (freq > nyquist) break;
                harmonics.push([freq, 1 / n]);
            }
            break;

        case "triangle":
            // Triangle wave: odd harmonics, amplitude 1/n^2.
            for (let n = 1; ; n += 2) {
                const freq = baseFreq * n;
                if (freq > nyquist) break;
                // The sign alternates, but for additive synthesis amplitude is what matters.
                harmonics.push([freq, 1 / (n * n)]);
            }
            break;

        case "sawtooth":
            // Sawtooth wave: all harmonics, amplitude 1/n.
            for (let n = 1; ; n += 1) {
                const freq = baseFreq * n;
                if (freq > nyquist) break;
                harmonics.push([freq, 1 / n]);
            }
            break;

        default:
            // Default to a sine wave if waveform is unknown.
            harmonics.push([baseFreq, 1]);
            break;
    }

    return harmonics;
};

File: wasm/src/createWaveform.cpp
#include "dsp.hpp"
#include <string>    // For std::string
#include <vector>    // For std::vector
#include <algorithm>

using namespace std;

vector<vector<double>> createWaveform(const string& type,double baseFreq) {
    
    vector<vector<double>> harmonics;

    const int MAX_HARMONICS_COUNT = 15;

    if (type == "sine") {
        harmonics.push_back({baseFreq, 1.0});
    } else if (type == "square") {
        // A square wave consists of odd harmonics. The amplitude of the nth harmonic is 1/n.	
        for (int n = 1; n <= MAX_HARMONICS_COUNT * 2 - 1; n += 2) {
            harmonics.push_back({static_cast<double>(n) * baseFreq, 1.0 / static_cast<double>(n)});
        }
    } else if (type == "sawtooth") {
        // A sawtooth wave consists of all integer harmonics. The amplitude of the nth harmonic is 1/n.
        for (int n = 1; n <= MAX_HARMONICS_COUNT; ++n) { // Iterate through all integers
            harmonics.push_back({static_cast<double>(n) * baseFreq, 1.0 / static_cast<double>(n)});
        }
    } else if (type == "triangle") {
        // A triangle wave consists of odd harmonics. The amplitude of the nth harmonic is 1/(n^2).
        for (int n = 1; n <= MAX_HARMONICS_COUNT * 2 - 1; n += 2) { // Iterate through odd numbers
            harmonics.push_back({static_cast<double>(n) * baseFreq, 1.0 / (static_cast<double>(n) * static_cast<double>(n))});
        }
    }
    // If an unknown waveform type is provided, an empty vector will be returned.

    return harmonics;
}
